lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -3325,27 +3376,44 @@ void AdjointEmitter::rematerializeOriginalInstruction(SILInstruction *inst) {
   rematCloner.visit(inst);
 }
 
-// TODO: What if value was not loadable? Consider removing this functon entirely
-// and force indirect for all cases for the ease of implementation.
-SILValue AdjointEmitter::materializeAdjoint(AdjointValue value,
-                                            SILLocation loc) {
-  auto valueBuf = getBuilder().createAllocStack(loc, value.getType());
-  SWIFT_DEFER { getBuilder().createDeallocStack(loc, valueBuf); };
-  materializeAdjointIndirect(value, valueBuf);
-  auto access = getBuilder().createBeginAccess(
-      loc, valueBuf, SILAccessKind::Read, SILAccessEnforcement::Static,
-      /*noNestedConflict*/ true,
-      /*fromBuiltin*/ false);
-  auto loq = getBufferLOQ(value.getType().getASTType(), getAdjoint());
-  auto val = getBuilder().createLoad(loc, valueBuf, loq);
-  getBuilder().createEndAccess(loc, access, /*aborted*/ false);
-  return val;
+SILValue AdjointEmitter::materializeAdjointDirect(AdjointValue val,
+                                                  SILLocation loc) {
+  auto &builder = getBuilder();
+  auto &ctx = getContext();
+  switch (val.getKind()) {
+  case AdjointValue::Kind::Zero: {
+    if (auto tupleTy = val.getType().getAs<TupleType>()) {
+      auto elts = map<SmallVector<SILValue, 8>>(val.getTupleElements(),
+          [&](AdjointValue eltVal) {
+            return materializeAdjointDirect(eltVal, loc);
+          });
+      return builder.createTuple(loc, elts);
+    }
+    return createScalarValueDirect(
+        0, val.getType().getASTType(), loc, builder, ctx);
```

Comment:
It should, since adjoint values are only ever created for values of a types that satisfies those requirements.  There are a bunch of assertions in `createScalarValueIndirect`.